### PR TITLE
Implement basics of consensus module

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: consensus/src/lib.rs [hunk 1]
```diff
@@ -1,23 +1,49 @@
...
 #[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct ConsensusState {
-    pub non_nil_votes: HashMap<Hash256, HashSet<PublicKey>>,
-    pub nil_votes: HashMap<Hash256, HashSet<PublicKey>>,
-    pub height: BlockHeight,
-    pub round: ConsensusRound,
+pub struct State {
+    pub consensus_state: ConsensusState,
+    /// The set of messages that have been already updated to the Vetomint state machine.
+    pub updated_message: BTreeSet<Hash256>,
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:24]:
```suggestion
    pub updated_messages: BTreeSet<Hash256>,
```

File: consensus/src/lib.rs [hunk 2]
```diff
@@ -1,23 +1,49 @@
...
+    /// The set of the block hashes that have been verified.
+    pub verified_block_hash: Vec<Hash256>,
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:26]:
```suggestion
    pub verified_block_hashes: Vec<Hash256>,
```
Now I'm kind of confused whether it is better to use plurals..

File: consensus/src/lib.rs [hunk 3]
```diff
@@ -1,23 +1,49 @@
...
+    /// The validator set eligible for this block
+    pub validator_set: Vec<(PublicKey, VotingPower)>,
+    /// If this node is a particiapnt, the index of this node.
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:29]:
```suggestion
    /// If this node is a participant, the index of this node.
```

File: consensus/src/lib.rs [hunk 4]
```diff
@@ -1,23 +1,49 @@
...
+    pub this_node_index: Option<usize>,
+    /// If true, this every operation on this instance will fail; the user must
+    /// run `create()` to create a new instance.
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:32]:
```suggestion
    /// If true, any operation on this instance will fail; the user must
    /// run `create()` to create a new instance.
```

File: consensus/src/lib.rs [hunk 5]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
+    /// if it is guaranteed that the lock is not held for a long time.
+    verified_block_hash: Arc<parking_lot::RwLock<BTreeSet<Hash256>>>,
+    validator_set: BTreeSet<PublicKey>,
+}
+
+impl MessageFilter for ConsensusMessageFilter {
+    fn filter(&self, message: &Message) -> Result<(), String> {
+        serde_json::from_str::<ConsensusMessage>(message.data()).map_err(|e| e.to_string())?;
+        if !self.validator_set.contains(message.signature().signer()) {
+            return Err("The signer is not in the validator set".to_string());
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:69]:
```suggestion
            return Err("the signer is not in the validator set".to_string());
```

File: consensus/src/lib.rs [hunk 6]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
+        }
+        if self
+            .verified_block_hash
+            .read()
+            .contains(&message.to_hash256())
+        {
+            Ok(())
+        } else {
+            Err("The block hash is not validated yet.".to_string())
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:78]:
```suggestion
            Err("the block hash is not validated yet.".to_string())
```

File: consensus/src/lib.rs [hunk 7]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
+        timestamp: Timestamp,
+        consensus_params: ConsensusParams,
     ) -> Result<(), Error> {
-        unimplemented!()
+        let this_node_index = this_node_key
+            .map(|key| {
+                validator_set
+                    .iter()
+                    .position(|(pk, _)| *pk == validator_set[key].0)
+                    .ok_or_else(|| anyhow::anyhow!("The validator set does not contain this node."))
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:108]:
```suggestion
                    .ok_or_else(|| anyhow::anyhow!("the validator set does not contain this node."))
```

File: consensus/src/lib.rs [hunk 8]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
+            })
+            .transpose()?;
+        let height_info = HeightInfo {
+            validators: validator_set.iter().map(|(_, v)| *v).collect(),
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:112]: why is it named validators? the proper meaning of the code is total voting power i guess.

File: consensus/src/lib.rs [hunk 9]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
+            .await?;
+        let state: State = serde_json::from_str(&state)?;
+        if let Some(index) = state.this_node_index {
+            if state.validator_set[index].0
+                != this_node_key
+                    .as_ref()
+                    .ok_or_else(|| anyhow::anyhow!("private key is required"))?
+                    .public_key()
+            {
+                anyhow::bail!("the private key does not match");
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:150]:
```suggestion
                anyhow::bail!("private key does not match");
```

File: consensus/src/lib.rs [hunk 10]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
+        }
+        dms.set_filter(Arc::new(ConsensusMessageFilter {
+            verified_block_hash: Arc::clone(&verified_block_hash),
+            validator_set: state
+                .validator_set
+                .iter()
+                .map(|(pk, _)| pk.clone())
+                .collect(),
+        }));
+
```
Comment (TomTaehoonKim) [inline consensus/src/lib.rs:161]:
```suggestion
```

File: consensus/src/lib.rs [hunk 11]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
     NonNilPreComitted(ConsensusRound, Hash256, Timestamp),
     NilPreVoted(ConsensusRound, Timestamp),
     NilPreComitted(ConsensusRound, Timestamp),
-    Finalized(Timestamp),
+    Finalized(Hash256, Timestamp),
+}
+
+pub struct ConsensusMessageFilter {
+    /// Note that it is even DESIRABLE to use a synchronous lock in the async context,
```
Comment (YoonhoKim0527) [inline consensus/src/lib.rs:59]:
```suggestion
    /// Note that it is even DESIRABLE to use a synchronous lock in the async context
```

**Review by YoonhoKim0527 — COMMENTED** [review]
If there is a comma, the sentence below looks like a new sentence.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: consensus/src/lib.rs [hunk 12]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
     ) -> Result<(), Error> {
-        unimplemented!()
+        let this_node_index = this_node_key
+            .map(|key| {
+                validator_set
+                    .iter()
+                    .position(|(pk, _)| *pk == validator_set[key].0)
+                    .ok_or_else(|| anyhow::anyhow!("The validator set does not contain this node."))
+            })
+            .transpose()?;
```
Comment (lemonholic) [inline consensus/src/lib.rs:110]: This is equal to `let this_node_index = this_node_key;`.
Did you mean
```rust
let this_node_index = this_node_key
.map(|key| {
    validator_set
        .iter()
        .position(|(pk, _)| *pk == key)
        .ok_or_else(|| anyhow::anyhow!("the validator set does not contain this node."))
})
.transpose()?;
```
with `this_node_key: Option<PublicKey>`?

File: consensus/src/lib.rs [hunk 13]
```diff
@@ -1,23 +1,49 @@
...
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum ConsensusMessage {
+    Proposal(Hash256),
+    NonNilPreVoted(
+        ConsensusRound,
+        /// The hash of the voted block
+        Hash256,
+    ),
+    NonNilPreComitted(ConsensusRound, Hash256),
```
Comment (lemonholic) [inline consensus/src/lib.rs:44]: How about just removing `NonNil` prefix?

File: consensus/src/lib.rs [hunk 14]
```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
...
     NilPreVoted(ConsensusRound, Timestamp),
     NilPreComitted(ConsensusRound, Timestamp),
-    Finalized(Timestamp),
+    Finalized(Hash256, Timestamp),
+}
+
+pub struct ConsensusMessageFilter {
+    /// Note that it is even DESIRABLE to use a synchronous lock in the async context,
+    /// if it is guaranteed that the lock is not held for a long time.
+    verified_block_hash: Arc<parking_lot::RwLock<BTreeSet<Hash256>>>,
```
Comment (lemonholic) [inline consensus/src/lib.rs:61]: verified_block_hashes?

Comment (junha1) [reply consensus/src/lib.rs:61]: fixed

File: consensus/src/lib.rs (hunk 12 above)
Comment (junha1) [reply consensus/src/lib.rs:110]: god damn copilot tricked me. fixed.

File: consensus/src/lib.rs (hunk 13 above)
Comment (lemonholic) [reply consensus/src/lib.rs:44]: got it

Comment (lemonholic) [reply consensus/src/lib.rs:44]: #117

File: consensus/src/lib.rs [hunk 15]
```diff
@@ -26,28 +52,126 @@ pub enum ProgressResult {
...
+        }
+        if self
+            .verified_block_hashes
+            .read()
+            .contains(&message.to_hash256())
+        {
+            Ok(())
+        } else {
+            Err("the block hash is not validated yet.".to_string())
+        }
```
Comment (lemonholic) [inline consensus/src/lib.rs:79]: How could we validate a block even before receiving it?

Comment (lemonholic) [reply consensus/src/lib.rs:79]: In addition, the error message in L78 says `not validated` while `verify` is used throughout the code.

Comment (lemonholic) [reply consensus/src/lib.rs:79]: What do you mean by "verify"? Is it the same with validating a block?

Comment (junha1) [reply consensus/src/lib.rs:79]: 1. Block proposals IS NOT propagated via the consensus module (consensus module is only about their hashes). Proposals are handled by the Repository module, so we can expect that a block proposal is received in the repository first, and once verified, the node will register it to the consensus module.
2. I mistakenly use the two words "verify" and "validate" interchangeably. I believe all cryptographic things must be referred as 'verify' EXCEPT the word 'validator' which is used by an old convention. Fixed it.

Comment (lemonholic) [reply consensus/src/lib.rs]: I thought the CSV module does not validate(word in blockchain context) blocks. Now I see the meaning of it, I think.

**Review by lemonholic — APPROVED** [review]
Good to go

**Pull request merged**